File: BM23_二叉树的前序遍历/python_solutions/stack_solution.py
from typing import List
from tree_node import TreeNode

#
# 给你二叉树的根节点 root ，返回它节点值的 前序 遍历。
# 
# 数据范围：二叉树的节点数量满足 1≤n≤100，二叉树节点的值满足 1≤val≤100，树的各节点的值各不相同
#
# @param root TreeNode类 
# @return int整型一维数组
#
class StackSolution:
    
    def preorderTraversal(self , root: TreeNode) -> List[int]:
        if root == None: 
            return []
        
        result = []
        stack = []
        stack.append(root)

        while len(stack):
            topNode = stack.pop()
            if topNode == None:
                continue

            result.append(topNode.val)
            stack.append(topNode.right)
            stack.append(topNode.left)

        return result

File: BM23_二叉树的前序遍历/java_solutions/StackSolution.java
package BM23_二叉树的前序遍历.java_solutions;

import java.util.ArrayList;
import java.util.Stack;

public class StackSolution {
    /**
     * @param root TreeNode类
     * @return int整型一维数组
     */
    public int[] preorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<Integer>();
        Stack<TreeNode> stack = new Stack<>();

        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode topNode = stack.pop();
            if (topNode == null)
                continue;
            
            result.add(topNode.val);
            stack.push(topNode.right);
            stack.push(topNode.left);
        }

        return convertToIntArray(result);
    }

    private int[] convertToIntArray(ArrayList<Integer> list) {
        int[] arr = new int[list.size()];
        for(int i = 0; i < list.size();i++){
            arr[i] = list.get(i);
        }
        return arr;
    }
}
